### PR TITLE
feat(gossipsub): allow compiling for WASM

### PR DESCRIPTION
**Pull request by onur-ozkan**
## Description

This modification removes deprecated dependency `wasm_timer` and enables wasm compatibility on the gossibsup protocol by simply substituting the `wasm_timer::Instant` with `instant::Instant`(which supports `fn checked_add`) and `wasm_timer::Interval` with `futures_ticker::Ticker`.

## Notes

~In my opinion, `wasm-timer` should be removed completely from the dependency tree since it's quite bad that relying on the dependency which is deprecated and no longer maintained. `instant` is also not so up-to-date dependency, but at very least it works today. But it might be broken someday as well if we make a change on standard library on rustc. Writing simple wrapper in the libp2p tree would be the most stable solution for this project since you can have complete control over it and update it whenever standard library adds or changes something.~

### CODE REVIEW

File: protocols/gossipsub/Cargo.toml
Comment (thomaseizinger) [inline protocols/gossipsub/Cargo.toml]: Is this the recommended way of enabling these features? I remember that in the past, we added a `wasm_bindgen` feature and forwarded that.

Users can then enable that conditionally on the target if they'd like.

**Comment by thomaseizinger** [comment]
> In my opinion, `wasm-timer` should be removed completely from the dependency tree since it's quite bad that relying on the dependency which is deprecated and no longer maintained.

Agreed! Are you happy to send a PR for the remaining usages? (I haven't checked whether there actually are any.)

**Review by thomaseizinger — COMMENTED** [review]
Thank you! One question.

File: protocols/gossipsub/Cargo.toml
Comment (onur-ozkan) [reply protocols/gossipsub/Cargo.toml]: > I remember that in the past, we added a wasm_bindgen feature and forwarded that.

That makes more sense. Will change it today.

**Comment by onur-ozkan** [comment]
> Are you happy to send a PR for the remaining usages?

Certainly, I will check/work on that matter during the upcoming weekend.

File: protocols/gossipsub/Cargo.toml
Comment (onur-ozkan) [reply protocols/gossipsub/Cargo.toml]: Added `wasm-bindgen` feature to `gossipsub` which handles the wasm dependency features

https://github.com/libp2p/rust-libp2p/blob/9d1a36f97c1e6d182c2d83a4c6624b6d7ae04478/protocols/gossipsub/Cargo.toml#L13-L14

File: protocols/gossipsub/Cargo.toml
Comment (thomaseizinger) [inline protocols/gossipsub/Cargo.toml:14]: We have a `wasm-bindgen` feature in the `libp2p` meta crate. Can you activate this from there please?

Also, you can remove the `getrandom` dependency here because we already activate that one in the `libp2p` meta-crate!

File: protocols/gossipsub/src/peer_score.rs
Comment (thomaseizinger) [inline protocols/gossipsub/src/peer_score.rs:32]: Can we remove the `wasm_timer` dependency from this crate now?

**Review by thomaseizinger — CHANGES_REQUESTED** [review]
Two more comments, otherwise LGTM :)

File: protocols/gossipsub/src/peer_score.rs
Comment (onur-ozkan) [reply protocols/gossipsub/src/peer_score.rs:32]: This requires additional work which I will check later. We can't replace all the `wasm_timer::Instant` with `instant::Instant` in gossipsub because of the type conflicts in some of the gossipsub modules.

File: protocols/gossipsub/Cargo.toml
Comment (onur-ozkan) [reply protocols/gossipsub/Cargo.toml:14]: > We have a wasm-bindgen feature in the libp2p meta crate. Can you activate this from there please?

Done.

> Also, you can remove the getrandom dependency here because we already activate that one in the libp2p meta-crate!

I can't enable it's feature if I remove it.

Comment (thomaseizinger) [reply protocols/gossipsub/Cargo.toml:14]: > > Also, you can remove the getrandom dependency here because we already activate that one in the libp2p meta-crate!
> 
> I can't enable it's feature if I remove it.

The meta-crate enables it for you! :)

https://github.com/libp2p/rust-libp2p/blob/master/libp2p/Cargo.toml#L83

File: protocols/gossipsub/src/peer_score.rs
Comment (thomaseizinger) [reply protocols/gossipsub/src/peer_score.rs:32]: > This requires additional work which I will check later. We can't replace all the `wasm_timer::Instant` with `instant::Instant` in gossipsub because of the type conflicts in some of the gossipsub modules.

Okay, is this PR by itself functional then? Unfortunately we don't have an wasm _runtime_ tests yet.

File: protocols/gossipsub/Cargo.toml
Comment (onur-ozkan) [reply protocols/gossipsub/Cargo.toml:14]: I think that will not work when we use gossipsub directly as a crate. Am I missing something?

Comment (onur-ozkan) [reply protocols/gossipsub/Cargo.toml:14]: e.g,

```toml
libp2p-gossipsub = { git = "https://github.com/ozkanonur/rust-libp2p.git", branch = "t", features = ["wasm-bindgen"] }
```

Because of https://github.com/ozkanonur/rust-libp2p/commit/88b39dc3fab80004e524f65ab19bd782ac725a9d this change, it will not compile due to missing `js` feature of getrandom.

File: protocols/gossipsub/src/peer_score.rs
Comment (onur-ozkan) [reply protocols/gossipsub/src/peer_score.rs:32]: > is this PR by itself functional then?

I think it really is. Gossipsub can't even finish the wasm compilation with the current versions. Which makes people to fork `gossipsub` and `wasm_timer` and use that modified ones instead of upstream gossipsub.

> Unfortunately we don't have an wasm runtime tests yet.

`instant::Instant` and `wasm_timer::Instant` are nearly identical. There shouldn't be any problem for non-wasm platforms since `instant` re-exports std library just like wasm_timer for non-wasm platforms. The difference is `instant::Instant` has more functions implemented from `std::time::Instant`.

File: protocols/gossipsub/Cargo.toml
Comment (onur-ozkan) [reply protocols/gossipsub/Cargo.toml:14]: If `libp2p-gossipsub` must be pulled with `libp2p` meta-crate together, then I can do that. I can remove `features` section in `libp2p-gossipsub`, in fact I can even remove `getrandom` and `instant` dependencies in `libp2p-gossipsub` because they already comes with the meta-crate. But if `libp2p` is missing in the dependency tree, then `gossipsub` will not be able to compile successfully. Both fine for me, I leave this up to you:)

Comment (thomaseizinger) [reply protocols/gossipsub/Cargo.toml:14]: Perhaps the `wasm-bindgen` feature in `libp2p` is also wrong but the way I'd see it is that depending on `getrandom` and activating `js` is something that users can always do themselves. We don't need to forward that feature-flag for them.

I don't know to be honest :D

At some point we will have to sit down and work out, what the best way is to handle these features. Every crate seems to solve WASM-compatibility in a different way so it is unclear what the correct way is.

Comment (onur-ozkan) [reply protocols/gossipsub/Cargo.toml:14]: > Perhaps the wasm-bindgen feature in libp2p is also wrong

Yes, I have noticed some of the issues as well. However, I didn't want to make extensive changes in this pull request. Perhaps I will address them in a separate one at a later time.

**Comment by thomaseizinger** [comment]
Thanks! I've pushed a few follow-up commits. Most notable, the crate is now included in our WASM build on CI!

If we are already working on WASM compatibility, I'd be great if we could also get rid of `wasm_timer` in this PR. Let me know if you are up for that. I think replacing `Interval` with `futures_timer::Delay` shouldn't be too difficult.

**Comment by onur-ozkan** [comment]
> Most notable, the crate is now included in our WASM build on CI!

Awesome :)

I believed that shipping this would greatly benefit many people, so I directly opened the pull request for it. But I am also completely okay with replacing `wasm_timer` entirely in this PR. Since this PR is a blocker for one of our project, I will probably do it in couple days.

**Comment by thomaseizinger** [comment]
> > Most notable, the crate is now included in our WASM build on CI!
> 
> Awesome :)
> 
> I believed that shipping this would greatly benefit many people, so I directly opened the pull request for it. But I am also completely okay with replacing `wasm_timer` entirely in this PR. Since this PR is a blocker for one of our project, I will probably do it in couple days.

We will make a new release soon but it is likely a week or so away. This PR should easily make it in even if it takes a few more days before we can land it :)

**Comment by onur-ozkan** [comment]
I decided to use `futures_ticker::Ticker` as a replacement of `wasm_timer::Interval`. The reason I did that is that crate already using the dependencies we have in libp2p dependency tree. It implements the ticker functionality on top of `futures_timer`.

`futures_timer` has an open issue about merging `futures_ticker`. See https://github.com/async-rs/futures-timer/issues/60

Also, it's worth to see this comment about `futures_timer` and `futures_ticker` https://github.com/libp2p/rust-libp2p/issues/2497#issuecomment-1038923700

If `futures_ticker` will be merged into `futures_timer`, we can use `futures_timer` easily in libp2p by simply replacing a few keywords in the codebase.

Let me know if this is acceptable or not :)

**Review by thomaseizinger — DISMISSED** [review]
Awesome!

I guess this means we can close https://github.com/libp2p/rust-libp2p/issues/2497 with this PR then?

**Comment by thomaseizinger** [comment]
cc @mxinden @AgeManning In https://github.com/libp2p/rust-libp2p/issues/2497, there was talk that the only remaining work would be to (re?) enable WASM support for gossipsub. This PR achieves this by means of https://github.com/antifuchs/futures-ticker.

**Comment by onur-ozkan** [comment]
> I guess this means we can close https://github.com/libp2p/rust-libp2p/issues/2497 with this PR then?

I think so.

Another way of solving this would be porting `futures_ticker` implementation in gossipsub. The implementation is super simple(~60 lines of code), and you can update it's dependencies(futures_timer, futures and instant) whenever you want. When you have those 3 dependencies updated in libp2p dependency tree, you will have multiple versions of the same dependency just because of `futures_ticker`.

**Comment by thomaseizinger** [comment]
> > I guess this means we can close #2497 with this PR then?
> 
> I think so.
> 
> Another way of solving this would be porting `futures_ticker` implementation in gossipsub. The implementation is super simple(~60 lines of code), and you can update it's dependencies(futures_timer, futures and instant) whenever you want. When you have those 3 dependencies updated in libp2p dependency tree, you will have multiple versions of the same dependency just because of `futures_ticker`.

If the dependencies aren't pinned, cargo should resolve them to the same version if they are semver compatible.https://github.com/libp2p/rust-libp2p/issues/2497 was created because we initially had our own interval implementation and it was buggy. I'd like to not go back to that (even without a buggy implemention :grinning: )

File: protocols/gossipsub/src/behaviour.rs
Comment (mxinden) [inline protocols/gossipsub/src/behaviour.rs]: Just to make sure this does not go unnoticed before merging.

**Review by mxinden — COMMENTED** [review]
Thanks for the work here. No objections from my end.

Note that `futures-tick` uses a "skip" style strategy, i.e. when one misses e.g. 2 ticks, only one is delivered. See also good documentation in [tokio's `Interval` implementation](https://docs.rs/tokio/latest/tokio/time/enum.MissedTickBehavior.html#variant.Skip).

Looking at `futures-tick`, I think there is room for improvement. See https://github.com/antifuchs/futures-ticker/pull/10.

File: protocols/gossipsub/src/behaviour.rs
Comment (onur-ozkan) [reply protocols/gossipsub/src/behaviour.rs]: https://github.com/antifuchs/futures-ticker/pull/11

Comment (thomaseizinger) [reply protocols/gossipsub/src/behaviour.rs]: Thanks to both of you!

**Comment by thomaseizinger** [comment]
> Note that `futures-tick` uses a "skip" style strategy, i.e. when one misses e.g. 2 ticks, only one is delivered. See also good documentation in [tokio's `Interval` implementation](https://docs.rs/tokio/latest/tokio/time/enum.MissedTickBehavior.html#variant.Skip).

I think that makes sense for us. This is used for heartbeats so I don't see a need for a burst strategy. Rather, performing the next heartbeat as soon as possible is important.

Do we agree?

**Comment by onur-ozkan** [comment]
> I think that makes sense for us. This is used for heartbeats so I don't see a need for a burst strategy. 

++

File: protocols/gossipsub/src/behaviour.rs
Comment (onur-ozkan) [reply protocols/gossipsub/src/behaviour.rs]: fixed with https://github.com/libp2p/rust-libp2p/pull/3973/commits/9a00bab0b712679d2ca0b0a96df28949b2f117c7

**Comment by mxinden** [comment]
> > Note that `futures-tick` uses a "skip" style strategy, i.e. when one misses e.g. 2 ticks, only one is delivered. See also good documentation in [tokio's `Interval` implementation](https://docs.rs/tokio/latest/tokio/time/enum.MissedTickBehavior.html#variant.Skip).
> 
> I think that makes sense for us. This is used for heartbeats so I don't see a need for a burst strategy. Rather, performing the next heartbeat as soon as possible is important.
> 
> Do we agree?

yes

File: protocols/gossipsub/src/behaviour.rs
Comment (mxinden) [reply protocols/gossipsub/src/behaviour.rs]: Thank you @ozkanonur.

**Review by mxinden — DISMISSED** [review]
Thank you for the work here!

**Comment by mxinden** [comment]
> and enables wasm compatibility on the gossibsup protocol

If you don't mind sharing, I would like to know what you are building with rust-libp2p and WASM @ozkanonur.

**Comment by AgeManning** [comment]
Sorry for the late input here. 

As you've all pointed out, we had significant issues because in the past because there was a buggy implementation of interval. It's very important that the heartbeat stays consistent. It appears the ticker approach behaves like what we would need, in that if the software is over-loaded heartbeats are skipped, but should still occur at regular intervals.

We (Lighthouse) have no affinity to wasm-timer. So changes look good, provided the heartbeat works as expected :)

**Review by mxinden — APPROVED** [review]

**Comment by onur-ozkan** [comment]
> > and enables wasm compatibility on the gossibsup protocol
> 
> If you don't mind sharing, I would like to know what you are building with rust-libp2p and WASM @ozkanonur.

We have a dex/swap protocol called https://github.com/KomodoPlatform/atomicDEX-API that utilizes libp2p for the network stack. Initially, it was written in C, but around 2019, we made the decision to rewrite it in Rust. As part of this rewrite, the network stack was migrated from nanomsg to libp2p, which involved creating custom patches for the libp2p library.

Today, I am rewriting the network layer of the project(https://github.com/KomodoPlatform/atomicDEX-API/issues/1756). This involves removing the forks and instead utilizing the original version of libp2p (the upstream). Throughout this process, I will likely come up with several more ideas and proposals to improve the libp2p upstream if I believe they would be beneficial.

**Comment by thomaseizinger** [comment]
> As you've all pointed out, we had significant issues because in the past because there was a buggy implementation of interval. It's very important that the heartbeat stays consistent. It appears the ticker approach behaves like what we would need, in that if the software is over-loaded heartbeats are skipped, but should still occur at regular intervals.

This sounds like we should have a test for that :)

I'll create an issue.

**Pull request merged**
